pkg/overlay/cache.go

```diff
@@ -107,51 +107,48 @@ func (o *Cache) Bootstrap(ctx context.Context) error {
 			return err
 		}
 	}
-	// called after kademlia is bootstrapped
-	// needs to take RoutingTable and start to persist it into the cache
-	// take bootstrap node
-	// get their route table
-	// loop through nodes in RT and get THEIR route table
-	// keep going forever and ever
-
-	// Other Possibilities: Randomly generate node ID's to ask for?
-
-	_, err = o.DHT.GetRoutingTable(ctx)
 
 	return err
 }
 
-// Refresh walks the network looking for new nodes and pings existing nodes to eliminate stale addresses
+// Refresh will run as often as configured and updates the cache db
+// with the current DHT. We currently do not penalize nodes that are
+// unresponsive, but should in the future.
 func (o *Cache) Refresh(ctx context.Context) error {
-	// iterate over all nodes
-	// compare responses to find new nodes
-	// listen for responses from existing nodes
-	// if no response from existing, then mark it as offline for time period
-	// if responds, it refreshes in DHT
-	_, rtErr := o.DHT.GetRoutingTable(ctx)
-
-	if rtErr != nil {
-		return rtErr
+	table, err := o.DHT.GetRoutingTable(ctx)
+	if err != nil {
+		zap.Error(OverlayError.New("Error getting routing table", err))
+		return err
 	}
 
-	_, err := o.DHT.GetNodes(ctx, "0", 128)
+	k := table.K()
+	nodes, err := o.DHT.GetNodes(ctx, "0", k)
```

Comment:
so this is getting the closest k nodes to "0". is the node id "0" or byte 0x30 (ascii code for the character 0)?

the kademlia paper says that refreshes should happen per bucket (you'll probably need to ask the DHT for what the buckets are), and in each bucket a random value should be chosen and a lookup should be performed. but perhaps that's the Walk method? what function from the paper is this refresh doing?

Reply:
I could write this to cycle through buckets instead of just nodes. We had discussed doing random node ID lookups as well before. That wouldn't be hard to add in right now either.  I'll take a look at that.